include/dali/core/cuda_stream_pool.h

```diff
@@ -0,0 +1,212 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_CUDA_STREAM_POOL_H_
+#define DALI_CORE_CUDA_STREAM_POOL_H_
+
+#include <cassert>
+#include <atomic>
+#include <vector>
+#include <utility>
+#include "dali/core/cuda_stream.h"
+#include "dali/core/spinlock.h"
+
+namespace dali {
+
+class CUDAStreamLease;
+
+class DLL_PUBLIC CUDAStreamPool {
+ public:
+  CUDAStreamPool();
+  ~CUDAStreamPool();
+
+  /**
+   * @brief Get a stream for given device.
+   *
+   * @param device_id   CUDA runtime API device ordinal. If negative, calling thread's
+   *                    current device is used.
+   *
+   * @return A CUDA stream wrapper object. If there were any streams in the pools, the
+   *         stream is taken from it, otherwise a new stream is created.
+   */
+  CUDAStreamLease Get(int device_id = -1);
+
+  /**
+   * @brief Place a stream for given device in the pool.
+   *
+   * @param device_id CUDA runtime API device ordinal of the device for which the stream was
+   *                  created. If negative, the device is obtained form the device context
+   *                  associated with the stream.
+   *
+   * @remarks It is an error to misstate the device_id. Placing a stream with improper deviceid
+   *          will render the stream pool unusable.
+   */
+  void Put(CUDAStream &&stream, int device_id = -1);
+
+  /**
+   * @brief Remove all streams currently in the pool and deletes auxiliary data structures.
+   *
+   * NOTE: Avoid using this function on the pool returned by `instance()`
+   */
+  void Purge();
+
+  /**
+   * @brief Returns a reference to the singleton instance.
+   *
+   * @remarks Using the singleton instance is only possible if:
+   * - the enclosing library is compiled and used as a shared object,
+   * - the enclosing library is static, but matching calls to Get and Put calls are
+   *   contained within one shared object.
+   */
+  static CUDAStreamPool &instance();
+
+ private:
+  friend class CUDAStreamLease;
+  friend class CUDAStreamPoolTest;
+
+  CUDAStream GetFromPool(int device_id);
+
+  struct StreamEntry {
+    StreamEntry() = default;
+    explicit StreamEntry(CUDAStream stream, StreamEntry *next = nullptr)
+    : stream(std::move(stream)), next(next) {}
+    CUDAStream stream;
+    StreamEntry *next = nullptr;
+  };
+
+  StreamEntry *unused_ = nullptr;
+  std::atomic_int lease_count_;
+
+  vector<StreamEntry *> dev_streams_;
+  spinlock lock_;
+
+  static StreamEntry *Pop(StreamEntry *&head) {
```

Comment:
Because it hides the element type. The only way to "pump" elements from one list to another is to create a one-element intermediate list. It's neither efficient nor readable.